src/builder.rs

```diff
@@ -0,0 +1,162 @@
+use std::str::FromStr;
+use std::{collections::HashMap, iter::IntoIterator};
+
+use crate::{config::Config, error, path::Expression, source::Source, value::Value};
+
+/// A configuration builder
+///
+/// It registers ordered sources of configuration to later build consistent [`Config`] from them.
+/// Configuration sources it defines are defaults, [`Source`]s and overrides.
+///
+/// Defaults are alaways loaded first and can be overwritten by any of two other sources.
+/// Overrides are always loaded last, thus cannot be overridden.
+/// Both can be only set explicitly key by key in code
+/// using [`set_default`](Self::set_default) or [`set_override`](Self::set_override).
+///
+/// An intermediate category, [`Source`], set groups of keys at once implicitly using data coming from external sources
+/// like files, environment variables or others that one implements. Defining a [`Source`] is as simple as implementing
+/// a trait for a struct.
+///
+/// Adding sources, setting defaults and overrides does not invoke any I/O nor builds a config.
+/// It happens on demand when [`build`](Self::build) (or its alternative) is called.
+/// Therefore all errors, related to any of the [`Source`] will only show up then.
+///
+/// # Examples
+///
+/// ```rust
+/// # use config::*;
+/// # use std::error::Error;
+/// # fn main() -> Result<(), Box<dyn Error>> {
+/// let mut builder = ConfigBuilder::default();
+///
+/// builder.set_default("default", "1")?;
+/// builder.add_source(File::new("config/settings", FileFormat::Json));
+/// builder.set_override("override", "1")?;
+///
+/// match builder.build() {
+///     Ok(config) => {
+///         // use your config
+///     },
+///     Err(e) => {
+///         // something went wrong
+///     }
+/// }
+/// # Ok(())
+/// # }
+/// ```
+///
+/// Calls can be chained as well
+/// ```rust
+/// # use std::error::Error;
+/// # use config::*;
+/// # fn main() -> Result<(), Box<dyn Error>> {
+/// let mut builder = ConfigBuilder::default();
+///
+/// builder
+///     .set_default("default", "1")?
+///     .add_source(File::new("config/settings", FileFormat::Json))
+///     .add_source(File::new("config/settings.prod", FileFormat::Json))
+///     .set_override("override", "1")?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug, Clone, Default)]
+pub struct ConfigBuilder {
+    defaults: HashMap<Expression, Value>,
+    overrides: HashMap<Expression, Value>,
+    sources: Vec<Box<dyn Source + Send + Sync>>,
+}
+
+impl ConfigBuilder {
+    /// Set a default `value` at `key`
+    ///
+    /// This value can be overwritten by any [`Source`] or override.
+    ///
+    /// # Errors
+    ///
+    /// Fails if `Expression::from_str(key)` fails.
+    pub fn set_default<S, T>(&mut self, key: S, value: T) -> error::Result<&mut ConfigBuilder>
```

Comment:
I'm thinking about why we're using the `&mut self` builder-pattern style here. Maybe `mut self` would be more appropriate (moving and returning `Self`), so we can chain the `build()` function properly?

Nothing that blocks this PR from landing though!

That would also make the `impl Source for [Box<dyn Source + Send + Sync>]` unnecessary I guess.

Reply:
Actually, I've used `&mut self`, because this is how it was used in `Config`.

But point is valid, consuming version would allow to chain things properly, I am going to change this PR. After all if it turns out in the future that `&mut self` version is required by users, it is "less breaking" to lessen required ownership (self -> &self, etc.) than the other way around.

It would not make this implementation you mention unnecessary as it is needed because there is `build_cloned` function that takes `&self`. It makes sense to have a cloned version especially when speaking about configuration that can change over time (etcd, consul, even files in rare circumstances) that you might want refresh periodically in your application.

Reply:
> files in rare circumstances

not even in rare ones: Think of a server software which should be able to pick up config changes without restarting!

> I am going to change this PR

Awesome!

Reply:
I am happy you agree about the two flavors of the build method. I did not want to sound too pushy on this as I remember your opposition to the idea in some earlier comments, but maybe I just recall wrongly. 

Code is changed, ConfigBuilder's methods are not consuming, except for one of two build methods. Feel free to merge as soon as you want!

Reply:
I did not change commit messages exactly as you wanted, but I hope it is fine the way it is.